lib/src/chunking.rs

```diff
@@ -364,78 +398,374 @@ impl Chunking {
-    r.extend(components.into_iter().map(|v| vec![v]));
-    // Add the tail
-    r.extend(by_src.into_values());
-    // And order the new list
-    sort_packing(&mut r);
-    // It's possible that merging components gave us enough space; if so
-    // we're done!
-    if r.len() <= bins.get() as usize {
+
+    println!("Creating new packing structure");
+
+    if before_processing_pkgs_len < bin_size.get() as usize {
+        components.into_iter().for_each(|pkg| r.push(vec![pkg]));
+        if before_processing_pkgs_len > 0 {
+            let new_pkgs_bin: Vec<&ObjectSourceMetaSized> = Vec::new();
+            r.push(new_pkgs_bin);
+        }
         return r;
     }
 
-    let last = (bins.get().checked_sub(1).unwrap()) as usize;
-    // The "tail" is components past our maximum.  For now, we simply group all of that together as a single unit.
-    if let Some(tail) = r.drain(last..).reduce(|mut a, b| {
-        a.extend(b.into_iter());
-        a
-    }) {
-        r.push(tail);
+    let mut max_freq_components: Vec<&ObjectSourceMetaSized> = Vec::new();
+    components.retain(|pkg| {
+        let retain: bool = pkg.meta.change_frequency != u32::MAX;
+        if !retain {
+            max_freq_components.push(pkg);
+        }
+        retain
+    });
+    let components_len_after_max_freq = components.len();
+    match components_len_after_max_freq {
+        0 => (),
+        _ => {
+            //Defining Limits of each bins
+            let limit_ls_bins = 1usize;
+            let limit_new_bins = 1usize;
+            let _limit_new_pkgs = 0usize;
+            let limit_max_frequency_pkgs = max_freq_components.len();
+            let limit_max_frequency_bins = if limit_max_frequency_pkgs > 0 {
+                1usize
+            } else {
+                0usize
+            };
+            let limit_hs_bins = (0.6
+                * (bin_size.get()
+                    - (limit_ls_bins + limit_new_bins + limit_max_frequency_bins) as u32)
+                    as f32)
+                .floor() as usize;
+            let limit_ms_bins = (bin_size.get()
+                - (limit_hs_bins + limit_ls_bins + limit_new_bins + limit_max_frequency_bins)
+                    as u32) as usize;
+            let partitions =
+                get_partitions_with_threshold(components, limit_hs_bins as usize, 2f64)
+                    .expect("Partitioning components into sets");
+
+            let limit_ls_pkgs = match partitions.get(LOW_PARTITION) {
+                Some(n) => n.len(),
+                None => 0usize,
+            };
+
+            let pkg_per_bin_ms: usize =
+                (components_len_after_max_freq - limit_hs_bins - limit_ls_pkgs)
+                    .checked_div(limit_ms_bins)
+                    .expect("number of bins should be >= 4");
+
+            //Bins assignment
+            for partition in partitions.keys() {
+                let pkgs = partitions.get(partition).expect("hashset");
```

Comment:
I think this is just
```suggestion
            for (partition, pkgs) in partitions.iter() {
```
right?